key/src/main/java/software/amazon/kms/key/UpdateHandler.java

```diff
@@ -58,9 +58,12 @@ public UpdateHandler(final ClientBuilder clientBuilder,
                 .updateKeyDescription(proxy, proxyClient, previousModel, model, callbackContext))
             .then(progress -> keyHandlerHelper
                 .updateKeyPolicy(proxy, proxyClient, previousModel, model, callbackContext))
-            .then(progress -> softFailAccessDenied(() -> keyHandlerHelper
-                .updateKeyTags(proxy, proxyClient, model, tags, callbackContext),
-                model, callbackContext))
+            .then(progress -> {
+                if (!tags.equals(request.getPreviousResourceTags())) {
```

Comment:
Why are we checking for tags equality here instead of directly calling KMS ?

Reply:
+1

> However, this change will have CFN not attempt any tag query/update if there are no modeled changes. This is different than soft failing access denied on ListResourceTags if there's no update being made. This difference can be seen if the customers' tags have drifted from what's modeled. Your current implementation will make no attempt to fix the drift, even if the permissions are in place. I don't know if that's what we want.

Yeah, This is not what we want. We should still attempt to fix drift. This should be achievable like so:
```
if (!tags.equals(request.getPreviousResourceTags())) {
// Customer is attempting to change tags, no soft fail
return keyHandlerHelper
                .updateKeyTags(proxy, proxyClient, model, tags, callbackContext);
} else { 
// Customer did not explicitly request a tag update, fix the drift if possible
return softFailAccessDenied(() -> keyHandlerHelper
                .updateKeyTags(proxy, proxyClient, model, tags, callbackContext));
}
```